File: 30-Day LeetCoding Challenge/C++/diameterOfBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    
public:
    pair<int, int> findDiameterAndDepth(TreeNode* node) {
        if(node == NULL)
            return {0,0};
        pair<int, int> left_node = findDiameterAndDepth(node->left);
        pair<int, int> right_node = findDiameterAndDepth(node->right);
        int longest_path = max({left_node.first, right_node.first, left_node.second + right_node.second});
        return {longest_path, max(left_node.second, right_node.second) + 1};
    }
    int diameterOfBinaryTree(TreeNode* root) {
        return findDiameterAndDepth(root).first;
    }
};

File: 30-Day LeetCoding Challenge/Python/diameterOfBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findDiameterAndDepth(self, root):
        if root == None:
            return [0, 0]
        left = self.findDiameterAndDepth(root.left)
        right = self.findDiameterAndDepth(root.right)
        longest_path = max(left[0], right[0], left[1] + right[1])
        depth = max(left[1], right[1]) + 1
        return [longest_path, depth]
    def diameterOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        ans = self.findDiameterAndDepth(root)
        return ans[0]
        